rust/src/mqtt/detect.rs

```diff
@@ -17,421 +17,1532 @@
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::UNSUBACK(ref unsuback) = msg.op {
+            if let Some(ref reason_codes) = unsuback.reason_codes {
+                for rc in reason_codes.iter() {
+                    if detect_match_uint(code, *rc) {
+                        return 1;
+                    }
+                }
+            }
+        }
+    }
     return 0;
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn rs_mqtt_tx_get_subscribe_topic(
-    tx: &MQTTTransaction, i: u32, buf: *mut *const u8, len: *mut u32,
-) -> u8 {
+static mut UNSUB_TOPIC_MATCH_LIMIT: isize = 100;
+static mut G_MQTT_UNSUB_TOPIC_BUFFER_ID: c_int = 0;
+static mut G_MQTT_TYPE_KW_ID: c_int = 0;
+static mut G_MQTT_TYPE_BUFFER_ID: c_int = 0;
+static mut SUB_TOPIC_MATCH_LIMIT: isize = 100;
+static mut G_MQTT_SUB_TOPIC_BUFFER_ID: c_int = 0;
+static mut G_MQTT_REASON_CODE_KW_ID: c_int = 0;
+static mut G_MQTT_REASON_CODE_BUFFER_ID: c_int = 0;
+static mut G_MQTT_QOS_KW_ID: c_int = 0;
+static mut G_MQTT_QOS_BUFFER_ID: c_int = 0;
+static mut G_MQTT_PUB_TOPIC_BUFFER_ID: c_int = 0;
+static mut G_MQTT_PUB_MSG_BUFFER_ID: c_int = 0;
+static mut G_MQTT_PROTOCOL_VERSION_KW_ID: c_int = 0;
+static mut G_MQTT_PROTOCOL_VERSION_BUFFER_ID: c_int = 0;
+static mut G_MQTT_FLAGS_KW_ID: c_int = 0;
+static mut G_MQTT_FLAGS_BUFFER_ID: c_int = 0;
+static mut G_MQTT_CONN_WILLTOPIC_BUFFER_ID: c_int = 0;
+static mut G_MQTT_CONN_WILLMSG_BUFFER_ID: c_int = 0;
+static mut G_MQTT_CONN_USERNAME_BUFFER_ID: c_int = 0;
+static mut G_MQTT_CONN_PROTOCOLSTRING_BUFFER_ID: c_int = 0;
+static mut G_MQTT_CONN_PASSWORD_BUFFER_ID: c_int = 0;
+static mut G_MQTT_CONN_CLIENTID_BUFFER_ID: c_int = 0;
+static mut G_MQTT_CONNACK_SESSIONPRESENT_KW_ID: c_int = 0;
+static mut G_MQTT_CONNACK_SESSIONPRESENT_BUFFER_ID: c_int = 0;
+static mut G_MQTT_CONN_FLAGS_KW_ID: c_int = 0;
+static mut G_MQTT_CONN_FLAGS_BUFFER_ID: c_int = 0;
+
+unsafe extern "C" fn unsub_topic_get_data(
+    tx: *const c_void, _flow_flags: u8, local_id: u32, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    let ml = unsafe { UNSUB_TOPIC_MATCH_LIMIT };
+    if ml > 0 && local_id >= ml as u32 {
+        return false;
+    }
     let mut offset = 0;
+    let tx = cast_pointer!(tx, MQTTTransaction);
     for msg in tx.msg.iter() {
-        if let MQTTOperation::SUBSCRIBE(ref subv) = msg.op {
-            if (i as usize) < subv.topics.len() + offset {
-                let topic = &subv.topics[(i as usize) - offset];
-                if !topic.topic_name.is_empty() {
-                    *len = topic.topic_name.len() as u32;
-                    *buf = topic.topic_name.as_ptr();
-                    return 1;
+        if let MQTTOperation::UNSUBSCRIBE(ref unsubv) = msg.op {
+            if (local_id as usize) < unsubv.topics.len() + offset {
+                let topic = &unsubv.topics[(local_id as usize) - offset];
+                if !topic.is_empty() {
```

Comment:
I wonder if we treat right the empty topic case here

Reply:
You mean, in that empty topics should be returned nevertheless? Well, the spec does not forbid zero length topic strings (e.g. in MQTT5 if the topic alias property is used) so we might just return length of 0 and a null pointer, right?

Reply:
yes, we should...